File: DSA/Assignment_5/del_char.c
#include <stdio.h>
#include <stdlib.h>
typedef struct Node
{
    int data;
    struct Node *prev;
    struct Node *next;
}Node;

Node *createNode(int data)
{
    Node *newNode=(Node *)malloc(sizeof(Node));
    newNode->data=data;
    newNode->prev=NULL;
    newNode->next=NULL;
}

void insert_ele(Node* *head, int data)
{
    Node *newNode=createNode(data);
    if(*head==NULL)
    {
        *head=newNode;
        return;
    }
    Node *temp=*head;
    while(temp->next!=NULL)
    temp=temp->next;
    temp->next=newNode;
    newNode->prev=temp;
}

void del(Node* *head, int ele)
{
    Node *temp=*head;
    while(temp)
    {
        if(temp->data==ele)
        break;
        temp=temp->next;
    }
    if(temp==NULL)
    {
        printf("Element not in list!");
        return;
    }
    else if(temp->next==NULL)
    {
        temp->prev->next=NULL;
        free(temp);
    }
    else
    {
        temp->prev->next=temp->next;
        temp->next->prev=temp->prev;
        free(temp);
    }
}

int main()
{
    int n, data;
    printf("Enter number of nodes:");
    scanf("%d",&n);
    Node *head=NULL;
    for(int i=0;i<n;i++)
    {
        printf("Enter element:");
        scanf("%d",&data);
        insert_ele(&head, data);
    }
    
    
    return 0;
}

File: DSA/Assignment_5/del_char.py
class Node:
    def __init__(self, data):
        self.data=data
        self.next=None
        self.prev=None

class DoublyLinkedList:
    def __init__(self):
        self.head=None

    def createNode(self, data):
        return Node(data)
    
    def insert_ele(self, data):
        newNode=self.createNode(data)
        if self.head is None:
            self.head=newNode
            return
        
        temp=self.head
        while temp.next:
            temp=temp.next
        temp.next=newNode
        newNode.prev=temp

    def print_list(self):
        if self.head is None:
            print("Empty List!")
            return
        
        temp=self.head
        while temp.next:
            print(temp.data,end=' ')
            temp=temp.next
        print()

    def del_ele(self, ele):
        if self.head is None:
            print("Empty List!")
            return
        
        temp=self.head
        while temp.next:
            if temp.data==ele:
                break
            temp=temp.next
        if temp.next==None:
            print("Element not in list!")
            return
        temp.prev.next=temp.next
        if temp.next:
            temp.next.prev=temp.prev

def main():
    dll=DoublyLinkedList()

    #Formation of Doubly Linked List
    n=int(input("Enter number of nodes:"))
    for _ in range(n):
        ele=int(input("Enter element:"))
        dll.insert_ele(ele)
    
    print("Original List:",end='')
    dll.print_list()

    #Removal of element from list
    ele=int(input("Enter element to delete:"))
    dll.del_ele(ele)

    print()